Compilation failure 'exceeds maximum object size'

When attempting to compile libgd from master on Fedora Rawhide Linux (gcc 16.1.1), I get the following:

```

In function 'gdMalloc',
    inlined from 'compress_row' at gd_bmp.c:952:39,
    inlined from 'bmp_write_pixels_8bit' at gd_bmp.c:814:22,
    inlined from 'bmp_write_pixels' at gd_bmp.c:657:10,
    inlined from '_gdImageBmpCtx.part.0' at gd_bmp.c:407:10:
gdhelpers.c:75:16: error: argument 1 range [18446744071562067968, 18446744073709551615] exceeds maximum object size 9223372036854775807 [-Werror=alloc-size-larger-than=]
   75 |         return malloc (size);
      |                ^
/usr/include/stdlib.h: In function '_gdImageBmpCtx.part.0':
/usr/include/stdlib.h:676:14: note: in a call to allocation function 'malloc' declared here
  676 | extern void *malloc (size_t __size) __THROW __attribute_malloc__
      |              ^
In function 'memcpy',
    inlined from 'compress_row' at gd_bmp.c:957:2,
    inlined from 'bmp_write_pixels_8bit' at gd_bmp.c:814:22,
    inlined from 'bmp_write_pixels' at gd_bmp.c:657:10,
    inlined from '_gdImageBmpCtx.part.0' at gd_bmp.c:407:10:
/usr/include/bits/string_fortified.h:29:10: error: '__builtin___memcpy_chk' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
   29 |   return __builtin___memcpy_chk (__dest, __src, __len,
      |          ^
lto1: all warnings being treated as errors
make[3]: *** [/tmp/cc6reo47.mk:2: /tmp/cchJGb5O.ltrans0.ltrans.o] Error 1
lto-wrapper: fatal error: make returned 2 exit status
compilation terminated.
```

--- a/src/gd_bmp.c
+++ b/src/gd_bmp.c
@@ -366,7 +366,7 @@ static int _gdImageBmpCtx(gdImagePtr im, gdIOCtxPtr out, int bpp, int compressio
 	int ret = 1;
 	gdImagePtr write_im = im;
 
-	if (im == NULL || out == NULL) {
+	if (im == NULL || out == NULL || gdImageSX(im) <= 0 || gdImageSY(im) <= 0) {
 		return 1;
 	}
 

--- a/src/gd_bmp.c
+++ b/src/gd_bmp.c
@@ -366,7 +366,17 @@ static int _gdImageBmpCtx(gdImagePtr im, gdIOCtxPtr out, int bpp, int compressio
 	int ret = 1;
 	gdImagePtr write_im = im;
 
-	if (im == NULL || out == NULL || gdImageSX(im) <= 0 || gdImageSY(im) <= 0) {
+	if (im == NULL || out == NULL) {
+		return 1;
+	}
+
+	if (gdImageSX(im) <= 0 || gdImageSY(im) <= 0) {
+		gd_error("image dimensions must be greater than 0");
+		return 1;
+	}
+
+	if (overflow2(gdImageSX(im), gdImageSY(im))) {
+		gd_error("image dimensions are too large");
 		return 1;
 	}
 

